lib/ae_mdw/db/sync/transaction.ex

```diff
@@ -52,35 +62,83 @@ defmodule AeMdw.Db.Sync.Transaction do
   def sync(from_height, to_height, txi) when from_height > to_height,
     do: txi
 
+  @spec transaction_mutations(
+          {Node.signed_tx(), Txs.txi()},
+          {Blocks.block_index(), Blocks.time(), Contract.grouped_events()},
+          boolean()
+        ) :: [Mutation.t()]
+  def transaction_mutations(
+        {signed_tx, txi},
+        {block_index, mb_time, mb_events} = tx_ctx,
+        inner_tx? \\ false
+      ) do
+    {mod, tx} = :aetx.specialize_callback(:aetx_sign.tx(signed_tx))
+    tx_hash = :aetx_sign.hash(signed_tx)
+    type = mod.type()
+    model_tx = Model.tx(index: txi, id: tx_hash, block_index: block_index, time: mb_time)
+
+    contract_events_mutation =
+      if type == :contract_call_tx do
+        ct_pk = :aect_call_tx.contract_pubkey(tx)
+        events = Map.get(mb_events, tx_hash, [])
+
+        ContractEventsMutation.new(ct_pk, events, txi)
+      end
+
+    inner_tx_mutations =
+      if type == :ga_meta_tx or type == :paying_for_tx do
+        inner_signed_tx = Sync.InnerTx.signed_tx(type, tx)
+        # indexes the inner with the txi from the wrapper/outer
+        transaction_mutations({inner_signed_tx, txi}, tx_ctx, true)
+      end
+
+    :ets.insert(:tx_sync_cache, {txi, model_tx})
+
+    [
+      WriteTxMutation.new(model_tx, type, txi, mb_time, inner_tx?),
+      WriteLinksMutation.new(type, tx, signed_tx, txi, tx_hash, block_index),
+      contract_events_mutation,
+      WriteFieldsMutation.new(type, tx, block_index, txi),
+      inner_tx_mutations
+    ]
+    |> Enum.reject(&is_nil/1)
+    |> List.flatten()
+  end
+
   ################################################################################
 
   defp sync_generation(height, txi) do
     {key_block, micro_blocks} = AE.Db.get_blocks(height)
+    kb_txi = (txi == 0 && -1) || txi
+    kb_header = :aec_blocks.to_key_header(key_block)
+    kb_hash = ok!(:aec_headers.hash_header(kb_header))
```

Comment:
Actually as a matter of practical result, in order to avoid the exceptions inside mnesia transactions, we should probably replace the banging functions behaviour of throwing/raising errors with an `:mnesia.abort`.

Reply:
A few things regarding these comment:
* The `ok!` is outside of a mnesia transaction, so there's no need to do any `:mnesia.abort`
* Any exception that is never supposed to happen shouldn't be handled gracefully, defensive code would just contaminate the codebase
* This PR does not fix exceptions that might happen inside the mnesia transaction that are never caught (those work just fine). This PR is about removing the exceptions that get caught inside the mnesia transaction.

Reply:
Yeah, I have marked the wrong `ok!`, my bad. I meant one inside the mutations that are executed inside the transaction like this:
https://github.com/aeternity/ae_mdw/blob/61b20faabaf9a83de9dd84650b2c0c6c65f31521/lib/ae_mdw/db/sync/name.ex#L116

The proposal would be to replace the use of banging functions like `cache_through_read!` with their non banging counterparts and the nil case should abort a transaction. It would be a complement to the same good idea of the PR, to decrease or eliminate as much as possible the exceptions inside the `:mnesia.transaction`.

Reply:
These kind of exceptions where the code would blow up if there's a problem are not problematic for mnesia. If the exception isn't captured, mnesia will abort the transaction and no strange errors will appear. 

The problem with exceptions in mnesia are those exceptions that are captured inside the transaction function